## 01 Import Excel file replacing blanks by null



### AIM: Read data from Excel file handling blank cells replacing them by proper null values NaN that can be processded by .dropna() method. Saving finally cleansed data to a new folder as a .csv file

## 1. Import Excel data into Python

- Import requried libraries. All three libraries below are available in the VE

In [2]:
import pandas as pd 
import os
import numpy as np

- Define paths to project and data folder

In [ ]:
wd = os.getcwd()
print(f'My working directory is:{wd}') 

My working directory is:/home/pablo/Documents/Pablo_ubuntu/Python_github_projects/ML-using-Python


In [5]:
project_folder = os.path.join('/home','pablo','Documents','Pablo_ubuntu','Python_github_projects','ML-using-Python')
os.chdir(project_folder)

In [6]:
wd = os.getcwd()
print(f'My working directory is:{wd}')

My working directory is:/home/pablo/Documents/Pablo_ubuntu/Python_github_projects/ML-using-Python


In [8]:
data_folder = os.path.join('/home','pablo','Documents','Pablo_ubuntu','Python_github_projects','ML-using-Python','data')

In [9]:
print(f'Data fodler path:{project_folder}')
print(f'Data fodler path:{data_folder}')

Data fodler path:/home/pablo/Documents/Pablo_ubuntu/Python_github_projects/ML-using-Python
Data fodler path:/home/pablo/Documents/Pablo_ubuntu/Python_github_projects/ML-using-Python/data


- Display data folder contents

In [10]:
data_folder_contents = os.listdir(data_folder)
print(f'Data folder contents:{data_folder_contents}')

Data folder contents:['monthly-milk-production-pounds.csv', 'winequality-white.csv', 'Monthly-AE-Time-Series-March-2025.xls', 'ONS_long_term_immigration_end2024.xlsx', 'AE_Attendances_TypeI_2010_2025.csv', '03_INE_Spain_natural_growh_births_deaths.xlsx', 'winequality-red.csv', 'INE Resident population country of birth Spain.xlsx', 'ONS_Figure_01_Long_term_emigration_immigration_net_migration.xlsx', 'Monthly-AE-Time-Series-January-2024.xls', 'wine_quality.zip', 'OCDE_countries_population_figures_1970_2022.csv', 'AE_Time_Series_Data_website.txt', '01 INE resident population by nationality Spain and CV 2002 2024.xlsx', 'AE_Attendances_2010_2024.csv', 'INE total and foreign population figures Spain.xlsx', 'Type_I_ATT_TEST.csv', 'all_wine_reset.csv', 'winequality.names', 'AE_Attendances_Aug2010_Mar_2025.csv', '02 INE Spain CV population stocks and flows 2002 2025.xlsx', 'Type_I_ATT_TRAIN.csv', 'ONS_Figure_2__Population_increase_in_mid-2023_was_driven_mostly_by_net_international_migration.xl

- Imporing Excel file "INE total and foreign population figures Spain.xlsx" file in \data folder.

In [17]:
data_raw = pd.read_excel(os.path.join('data','INE total and foreign population figures Spain.xlsx'),
                         sheet_name = 'INE_Total_population',
                         skiprows= 7,
                         usecols = 'A:B',
                         nrows = 212
                         )
data_raw.head() 

Total
0    1 de enero de 2024  48619695
1  1 de octubre de 2023  48486865
2    1 de julio de 2023  48320520
3    1 de abril de 2023  48205962
4    1 de enero de 2023  48085361

In [18]:
data_raw.head()
print(data_raw)

                              Total
0      1 de enero de 2024  48619695
1    1 de octubre de 2023  48486865
2      1 de julio de 2023  48320520
3      1 de abril de 2023  48205962
4      1 de enero de 2023  48085361
..                    ...       ...
207    1 de abril de 1972          
208    1 de enero de 1972  34408338
209  1 de octubre de 1971          
210    1 de julio de 1971  34216856
211    1 de abril de 1971          

[212 rows x 2 columns]


In [19]:
data_raw.columns

Index([' ', 'Total'], dtype='object')

- Remane default column names from imported Excel file into Python

In [21]:
data_raw.columns = ['Year','Total_population']
data_raw.head()

Year Total_population
0    1 de enero de 2024         48619695
1  1 de octubre de 2023         48486865
2    1 de julio de 2023         48320520
3    1 de abril de 2023         48205962
4    1 de enero de 2023         48085361

In [22]:
data_raw.tail() 

Year Total_population
207    1 de abril de 1972                 
208    1 de enero de 1972         34408338
209  1 de octubre de 1971                 
210    1 de julio de 1971         34216856
211    1 de abril de 1971

### 1. Remove NaN values from Dataframe

#### 1.1 First I need to deal with Empty cells present in the data

- First issue I need to tacke is that we don't have NaN values we have Empty values in the bottom cells in Total_population column

In [ ]:
data_raw.tail() 

Year Total_population
207    1 de abril de 1972                 
208    1 de enero de 1972         34408338
209  1 de octubre de 1971                 
210    1 de julio de 1971         34216856
211    1 de abril de 1971

- Once I deal with Empty Cells I will apply then the dropna() method on NaN cells. This is how dropna() method works: 
- The dropna() method removes any rows NaN values (missing values) in any column by default:

- I will be applying the dropna() method to remove Null values from previous dataset. This is how the dropna() works: 

- dropna() arguments: 
-   axis = 0 : drop rows (default)
-   axis = 1 : drop columns

- We take previous dataframe "data_raw" imported from Excel file and try to apply the dropna() function:

In [27]:
data_cleansed = data_raw.dropna(axis = 0, 
                subset = ['Total_population'],
                inplace = False)
data_cleansed.head()
data_cleansed.tail()

Year Total_population
207    1 de abril de 1972                 
208    1 de enero de 1972         34408338
209  1 de octubre de 1971                 
210    1 de julio de 1971         34216856
211    1 de abril de 1971

The reason why dropna() has not work in the above dataframe, is becasue those are NOT null cell values NaN, but instead they are Empty cells. So I need to fix first those empty cells using .replace() method.

#### 1.2 Dealing with empty cells using .replace() method 

- The np.isnan() function returns a boolean array indicating which elements in the input array are NaN

In [28]:
imported_data_na = data_cleansed.replace(r'^\s*$',np.nan, regex = True) 
print(imported_data_na)

                     Year  Total_population
0      1 de enero de 2024        48619695.0
1    1 de octubre de 2023        48486865.0
2      1 de julio de 2023        48320520.0
3      1 de abril de 2023        48205962.0
4      1 de enero de 2023        48085361.0
..                    ...               ...
207    1 de abril de 1972               NaN
208    1 de enero de 1972        34408338.0
209  1 de octubre de 1971               NaN
210    1 de julio de 1971        34216856.0
211    1 de abril de 1971               NaN

[212 rows x 2 columns]


/tmp/ipykernel_19538/2464472249.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  imported_data_na = data_cleansed.replace(r'^\s*$',np.nan, regex = True)


In [29]:
imported_data_na.tail()

Year  Total_population
207    1 de abril de 1972               NaN
208    1 de enero de 1972        34408338.0
209  1 de octubre de 1971               NaN
210    1 de julio de 1971        34216856.0
211    1 de abril de 1971               NaN

In [30]:
count_imported_data_na_rows = imported_data_na.shape[0]  # Gives number of rows
print(count_imported_data_na_rows)

212


#### 1.3 Then I can use dropna() method on imported_data_na dataframe

- Now that we have our data with Missing values in the "Total_population" column, I can use the .dropna() method

- dropna() arguments: 
-   axis = 0 : drop rows (default)
-   axis = 1 : drop columns
- subset = ['Year','Total_population'] : We want to subset just Year and Total_population from previous dataframe

In [ ]:
imported_data_clean_dropna = imported_data_na.dropna(axis = 0, 
                subset = ['Total_population'],
                inplace = False)
imported_data_clean_dropna.head()
imported_data_clean_dropna.tail()    

Year  Total_population
202  1 de julio de 1973        34980317.0
204  1 de enero de 1973        34800600.0
206  1 de julio de 1972        34595886.0
208  1 de enero de 1972        34408338.0
210  1 de julio de 1971        34216856.0

In [ ]:
count_row_clean_test = imported_data_clean_dropna.shape[0]  # Gives number of rows
print(count_row_clean_test)

112


The dropna() function has **removed** the cells containing **Missing value** values so the number of records has reduces. The total number of rows has gone down from 212 to 112 after we removed all Null values using dropna() method. 

## 2. Flag January rows to keep just those rows

We have from previous step is to rename previous dataframe where we removed the NaN null values using dropna() function.

Then we rename "imported_data_clean_dropna" dataframe into imported_data_clean dataframe.

In [28]:
imported_data_clean = imported_data_clean_dropna

In [29]:
imported_data_clean.head()
imported_data_clean.tail()

Year  Total_population
202  1 de julio de 1973        34980317.0
204  1 de enero de 1973        34800600.0
206  1 de julio de 1972        34595886.0
208  1 de enero de 1972        34408338.0
210  1 de julio de 1971        34216856.0

### 2.1 Creating a Flag using str.contrains() function 

- This is an example on how str.contains() function works

In [31]:
imported_data_clean_flag = imported_data_clean.copy()
imported_data_clean_flag.head()

Year  Total_population
0    1 de enero de 2024        48619695.0
1  1 de octubre de 2023        48486865.0
2    1 de julio de 2023        48320520.0
3    1 de abril de 2023        48205962.0
4    1 de enero de 2023        48085361.0

- This code below creates a new collumn called "Jan_Flag" that contains value "Y" for rows including "Enero" word in the "Year" column.

In [33]:
imported_data_clean_flag.loc[imported_data_clean_flag['Year'].str.contains('enero',case = False),'Jan_flag'] = 'Y'

We can see how the new 'Jan_flag' column has been created with 'Y' flag in "imported_data_clean" dataframe:

In [34]:
imported_data_clean_flag.head()

Year  Total_population Jan_flag
0    1 de enero de 2024        48619695.0        Y
1  1 de octubre de 2023        48486865.0      NaN
2    1 de julio de 2023        48320520.0      NaN
3    1 de abril de 2023        48205962.0      NaN
4    1 de enero de 2023        48085361.0        Y

### 2.2 Then we can subset rows with 'Y' Flag in them 

From the imported_population_clean_flag dataframe we will keep just records with Y values in Jan_flag column:

In [35]:
final_population_data = imported_data_clean_flag[imported_data_clean_flag.Jan_flag == 'Y']
final_population_data.head()
final_population_data.tail()


Year  Total_population Jan_flag
192  1 de enero de 1976        35946425.0        Y
196  1 de enero de 1975        35569375.0        Y
200  1 de enero de 1974        35177294.0        Y
204  1 de enero de 1973        34800600.0        Y
208  1 de enero de 1972        34408338.0        Y

In [36]:
df_rows = final_population_data
num_rows_final_population_data = len(df_rows.index)
print(f'Number of rows:{num_rows_final_population_data}')

Number of rows:53


- Another approach to subset rows suggested by Copilot

In [38]:
imported_data_clean.head()

Year  Total_population
0    1 de enero de 2024        48619695.0
1  1 de octubre de 2023        48486865.0
2    1 de julio de 2023        48320520.0
3    1 de abril de 2023        48205962.0
4    1 de enero de 2023        48085361.0

In [42]:
imported_data_clean_flag_test = imported_data_clean.copy()

In [43]:
final_population_data_copilot = imported_data_clean_flag_test.loc[imported_data_clean_flag_test['Year'].str.contains('enero', case = False)]
final_population_data_copilot.head()

Year  Total_population
0   1 de enero de 2024        48619695.0
4   1 de enero de 2023        48085361.0
8   1 de enero de 2022        47486727.0
12  1 de enero de 2021        47400798.0
16  1 de enero de 2020        47318050.0

## 3. Finally export DataFrame including just January figures data as .csv file

- In the previous step, we have created this "final_population_data" where we subset just January data: 

In [44]:
final_population_data.head()

Year  Total_population Jan_flag
0   1 de enero de 2024        48619695.0        Y
4   1 de enero de 2023        48085361.0        Y
8   1 de enero de 2022        47486727.0        Y
12  1 de enero de 2021        47400798.0        Y
16  1 de enero de 2020        47318050.0        Y

- First we will have to create a new folder called "data_cleansed" to store this cleansed output data.

In [45]:
if not os.path.exists('data_cleansed'):
    os.makedirs('data_cleansed')

The above code has created a new folder in our WD to store formatted data

- Then subset columns from fain_population_data" dataset, keeping just "year" and "Total_population" columns: 

In [46]:
final_population_data.head()

Year  Total_population Jan_flag
0   1 de enero de 2024        48619695.0        Y
4   1 de enero de 2023        48085361.0        Y
8   1 de enero de 2022        47486727.0        Y
12  1 de enero de 2021        47400798.0        Y
16  1 de enero de 2020        47318050.0        Y

In [47]:
final_population_data_output = final_population_data[['Year','Total_population']]
final_population_data_output.head()

Year  Total_population
0   1 de enero de 2024        48619695.0
4   1 de enero de 2023        48085361.0
8   1 de enero de 2022        47486727.0
12  1 de enero de 2021        47400798.0
16  1 de enero de 2020        47318050.0

- Finally save this "final_population_data_output" dataframe to "data_cleansed" folder as .csv file. Include index = False to remove index colum when exporting the final dataset as a .csv file.

In [48]:
final_population_data_output.to_csv(os.path.join('data_cleansed','final_population_data_cleansed.csv'),index = False)